File: spark/spark-streaming/src/main/scala/imooc/project/CourseClickDao.scala
package imooc.project

import imooc.hbase1.HBaseUtils
import org.apache.hadoop.hbase.client.Get
import org.apache.hadoop.hbase.util.Bytes

import scala.collection.mutable.ListBuffer

/**
  * @Description: 保存统计数据到Hbase的DAO
  * @author: HuangYn 
  * @date: 2018/11/1 20:45
  */
object CourseClickDao {

  val tableName = "imooc_course_clickcount"
  val cf = "info"
  val qualifier = "click_count"

  //保存每天到当前时间为止,课程的点击量
  def saveCourseClickCount(list: ListBuffer[CourseClickCount]): Unit = {
    if (list.nonEmpty && list.size > 1) {

      val table = HBaseUtils.getInstance.getTable("imooc_course_clickcount")
      list.foreach(ccc => {
        table.incrementColumnValue(
          ccc.day_course.getBytes(),
          cf.getBytes(),
          qualifier.getBytes(),
          ccc.click_count)
      })
    }
  }

  //保存每天到当前时间为止,从搜索引擎搜索进入课程的点击量
  def saveCourseSearchClickCount(list: ListBuffer[CourseSearchClickCount]): Unit = {
    if (list.nonEmpty && list.size > 1) {
      val table = HBaseUtils.getInstance().getTable("imooc_course_search_clickcount")
      list.foreach(c =>
        table.incrementColumnValue(Bytes.toBytes(c.day_from),
          Bytes.toBytes(cf),
          Bytes.toBytes(qualifier),
          c.click_count)
      )
    }
  }

  //查询数据,每天到当前时间为止,课程的点击量
  def getCourseClickCount(day: String, course: String): CourseClickCount = {
    val table = HBaseUtils.getInstance().getTable("imooc_course_clickcount")
    val result = table.get(new Get((day + "_" + course).getBytes()))
    val cell = result.getColumnCells(cf.getBytes(), qualifier.getBytes()).get(0)
    CourseClickCount(
      new String(cell.getRowArray, cell.getRowOffset, cell.getRowLength),
      Bytes.toLong(cell.getValueArray, cell.getValueOffset, cell.getValueLength)
    )
  }

  def getCourseSearchClickCount(day: String, from: String): CourseSearchClickCount = {
    val table = HBaseUtils.getInstance().getTable("imooc_course_search_clickcount")
    val result = table.get(new Get((day + "_" + from).getBytes()))
    val cell = result.getColumnCells(cf.getBytes(), qualifier.getBytes()).get(0)
    CourseSearchClickCount(
      new String(cell.getRowArray, cell.getRowOffset, cell.getRowLength),
      Bytes.toLong(cell.getValueArray, cell.getValueOffset, cell.getValueLength)
    )
  }


  def main(args: Array[String]): Unit = {

    //    val list = ListBuffer(CourseClickCount("20181111_666", 10))
    //    save(list)
    val ccc = getCourseClickCount("20181111", "666")
    print(s"${ccc.day_course} : ${ccc.click_count}")
  }

}

File: spark/spark-streaming-web/src/main/java/imooc/dao/CourseClickDao.java
package imooc.dao;

import imooc.domain.CourseClickCount;
import imooc.domain.CourseSearchClickCount;
import org.apache.hadoop.hbase.client.*;
import org.apache.hadoop.hbase.filter.Filter;
import org.apache.hadoop.hbase.filter.PrefixFilter;
import org.apache.hadoop.hbase.util.Bytes;
import org.springframework.stereotype.Repository;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @Description:
 * @author: HuangYn
 * @date: 2018/11/2 15:18
 */
@Repository
public class CourseClickDao {

    private String cf = "info";
    private String qualfier = "click_count";

    //查询某一天所有课程的访问记录
    public List<CourseClickCount> getDayCourseClickCount(String day) throws IOException {
        Table table = HBaseUtils.getInstance().getTable("imooc_course_clickcount");
        Filter dayFilter = new PrefixFilter(day.getBytes());
        Scan scan = new Scan();
        scan.setFilter(dayFilter);
        ResultScanner resultScanner = table.getScanner(scan);
        List<CourseClickCount> list = new ArrayList<>();
        resultScanner.forEach(result -> {
            CourseClickCount ccc = new CourseClickCount();
            ccc.setDayCourse(new String(result.getRow()));
            ccc.setClickCount(Bytes.toLong(result.getValue(cf.getBytes(), qualfier.getBytes())));
            list.add(ccc);
        });
        return list;
    }

    public CourseSearchClickCount getDayCourseSearchClickCount(String day, String from) {
        Table table = HBaseUtils.getInstance().getTable("imooc_course_search_clickcount");
        try {
            Result result = table.get(new Get(Bytes.toBytes(day + "_" + from)));
            long count = Bytes.toLong(result.getValue(Bytes.toBytes(cf), Bytes.toBytes(qualfier)));
            CourseSearchClickCount cscc = new CourseSearchClickCount();
            cscc.setDayFrom(day + "_" + from);
            cscc.setClickCount(count);
            return cscc;

        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    public static void main(String[] args) throws IOException {
        CourseClickDao dao = new CourseClickDao();
        List<CourseClickCount> list = dao.getDayCourseClickCount("20181102");
        list.forEach(System.out::println);
    }

}
